### PR TITLE
Add SCM integration link commands

### LINKED ISSUE #445: Add commands for repo link

### PR DESCRIPTION
**Pull request by brandon-welsch**
Close #445

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -156,6 +160,12 @@ COMMANDS:
      signup     Create your Scalingo account
      self       Get the logged in profile
      whoami     Get the logged in profile
+     
+   Integrations:
```
Comment (EtienneM) [inline README.md]: What are these modifications? Can't you make a PR with only the repo link commands?

File: README.md [hunk 2]
```diff
@@ -167,15 +177,26 @@ COMMANDS:
...
      migrations-create  Migrate an app to another region
      migrations         List all migrations linked to an app
      migrations-follow  Follow a running migration
 
-   Public SSH Keys:
-     keys         List your SSH public keys
-     keys-add     Add a public SSH key to deploy your apps
-     keys-remove  Remove a public SSH key
+   Repo Link:
+     repo-link                    Show repo link linked with your app
```
Comment (EtienneM) [inline README.md]: What is the difference between a repo link and an integration?

File: README.md (hunk 1 above)
Comment (brandon-welsch) [reply README.md]: I have merged #444 in this branch to reuse some functions integrated into #444

File: README.md (hunk 2 above)
Comment (brandon-welsch) [reply README.md]: A repo link is a link between a git repository, an integration and an app.
Do I have to explain it ?

File: repolink/manual_review_app.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+	}
+
+	c, err := config.ScalingoClient()
+	if err != nil {
+		return errgo.Notef(err, "fail to get Scalingo client")
+	}
+
+	err = c.SCMRepoLinkManualReviewApp(app, pullRequestID)
+	if err != nil {
+		return errgo.Notef(err, "fail to trigger manual review app")
```
Comment (EtienneM) [inline repolink/manual_review_app.go]:
```suggestion
		return errgo.Notef(err, "fail to manually create a review app")
```

File: repolink/manual_review_app.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+	}
+
+	io.Statusf("Manual review app deployment triggered for app '%s' with pull/merge request id '%s'.\n", app, pullRequestID)
```
Comment (EtienneM) [inline repolink/manual_review_app.go]:
```suggestion
	io.Statusf("Manual review app created for app '%s' with pull/merge request id '%s'.\n", app, pullRequestID)
```

File: repolink/show.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+	i, err := c.SCMIntegrationsShow(repoLink.AuthIntegrationUUID)
+	if err != nil {
+		return errgo.Notef(err, "fail to get integration of this repo link")
+	}
+
+	t := tablewriter.NewWriter(os.Stdout)
+	t.SetHeader([]string{
+		"App ID", "Integration ID", "Integration Name", "Owner", "Repo", "Branch", "Created At",
+		"Auto Deploy", "Review Apps Deploy", "Delete on close", "Delete on stale",
```
Comment (EtienneM) [inline repolink/show.go]:
```suggestion
		"Auto Deploy", "Review Apps Deploy", "Delete on Close", "Delete on Stale",
```

File: repolink/update.go [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+	if err != nil {
+		return errgo.Notef(err, "fail to get Scalingo client")
+	}
+
+	_, err = c.SCMRepoLinkUpdate(app, params)
+	if err != nil {
+		return errgo.Notef(err, "fail to update repo link")
+	}
+
+	io.Statusf("RepoLink has been updated for app '%s'.\n", app)
```
Comment (EtienneM) [inline repolink/update.go]:
```suggestion
	io.Statusf("Repo link has been updated for app '%s'.\n", app)
```

File: review_apps/show.go [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+	if err != nil {
+		return errgo.Notef(err, "fail to get Scalingo client")
+	}
+
+	reviewApps, err := c.SCMRepoLinkReviewApps(app)
+	if err != nil {
+		return errgo.Notef(err, "fail to get review apps for this app")
+	}
+	if len(reviewApps) == 0 {
+		io.Statusf("No review apps for '%s' app or app specified is not a parent app.\n", app)
```
Comment (EtienneM) [inline review_apps/show.go]:
```suggestion
		io.Statusf("No review app for '%s' or specified app is not a parent app.\n", app)
```

File: cmd/repo_link.go [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+
+		# See also 'repo-link', 'repo-link-update', 'repo-link-delete', 'repo-link-manual-deploy' and 'repo-link-manual-review-app'`,
+		Action: func(c *cli.Context) {
+			if len(c.Args()) != 2 {
+				cli.ShowCommandHelp(c, "repo-link-create")
+				return
+			}
+
+			currentApp := appdetect.CurrentApp(c)
+			integrationType := c.Args()[0]
```
Comment (EtienneM) [inline cmd/repo_link.go]: Should be `--type`

File: cmd/repo_link.go [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+			integrationURL := c.Args()[1]
```
Comment (EtienneM) [inline cmd/repo_link.go]: Should be `--url`
The type becomes optional if github.com or gitlab.com in the URL

File: cmd/repo_link.go [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+			if err != nil {
+				errorQuit(err)
+			}
+		},
+		BashComplete: func(c *cli.Context) {
+			_ = autocomplete.CmdFlagsAutoComplete(c, "repo-link-create")
+		},
+	}
+
+	repoLinkUpdateCommand = cli.Command{
```
Comment (EtienneM) [inline cmd/repo_link.go]: Add `--no-auto-deploy` to disable. The `CheckAndFillParams function is wrong. There is an issue if the flag is not mentioned.

**Comment by Soulou** [comment]
And `--no-auto-review-apps`?

**Comment by EtienneM** [comment]
```
$ scalingo integrations-add github
$ scalingo integrations-add github-enterprise --url https://github.mycorp.com --token <token>
$ scalingo integrations-add gitlab-self-hosted --url https://gitlab.scalingo.com --token <token>
$ scalingo integrations-add gitlab-self-hosted --url https://gitlab.mycorp.com --token <token>

# If SaaS GitHub integration already exists
$ scalingo -a my-app integration-links-create https://github.com/betagouv/civilsdeladefense
Branch to auto-deploy (empty to disable): prod
Automatically deploy pull requests [y/N]: N
→ Application has been linked to github repository betagouv/civilsdeladefense

# Without GitHub integration
$ scalingo -a my-app integration-links-create https://github.com/betagouv/civilsdeladefense
→ No github integration found, please go to https://auth.scalingo.com/oauth/github

# Like SaaS GitHub
$ scalingo -a my-app integration-links-create https://gitlab.com/betagouv/civilsdeladefense

$ scalingo -a my-app integration-links-create https://gitlab.scalingo.com/betagouv/civilsdeladefense
 ! No integration found for URL https://gitlab.scalingo.com
 ! Please run `scalingo integrations-add gitlab-self-hosted|github-enterprise --url https://gitlab.scalingo.com

# Integration already exists, same wizard as show before
$ scalingo -a my-app integration-links-create https://github.scalingo.com/betagouv/civilsdeladefense
```

```
scalingo -a my-app integration-links-update --no-auto-deploy
scalingo -a my-app integration-links-update --auto-deploy --branch production
scalingo -a my-app integration-links-delete
```
s/integration-links/integration-link/g

**Comment by EtienneM** [comment]
### With no integration set

1. GL
```
scalingo --app test-etienne integration-link-create https://gitlab.com/Scalingo/api
 !     No GitLab integration found, please follow this URL to add it:
 !     https://auth.st-sc.fr/users/gitlab/link
```
2. GH
```
scalingo --app test-etienne integration-link-create https://github.com/EtienneM/sample-go-martini
 !     No GitHub integration found, please follow this URL to add it:
 !     https://auth.st-sc.fr/users/github/link
```
3. GHE/GL self-hosted

```
scalingo --app test-etienne integration-link-create https://github.example.com/EtienneM/sample-go-martini
 !     No integration found for URL https://github.example.com/EtienneM/sample-go-martini.
 !     Please run the command:
 !     scalingo integrations-add gitlab-self-hosted|github-enterprise --url https://github.example.com --token <personal-access-token>
```

### With integration set

1. With flags
```
scalingo --app test-etienne integration-link-create https://github.com/EtienneM/sample-go-martini --auto-deploy --branch master
-----> Your app 'test-etienne' is linked to the repository https://github.com/EtienneM/sample-go-martini.
```

2. With interactive mode:

WIP

File: cmd/integration_link.go [hunk 11]
```diff
@@ -0,0 +1,429 @@
...
+		# See also 'integration-link', 'integration-link-update', 'integration-link-delete', 'integration-link-manual-deploy' and 'integration-link-manual-review-app'`,
+		Action: func(c *cli.Context) {
+			if len(c.Args()) != 1 {
+				cli.ShowCommandHelp(c, "integration-link-create")
+				return
+			}
+
+			currentApp := appdetect.CurrentApp(c)
+			integrationURL := c.Args()[0]
+			integrationType, err := scmintegrations.GetTypeFromURL(integrationURL)
```
Comment (johnsudaar) [inline cmd/integration_link.go]: Could you move all the "logic" to the scm_integrations package? In this package all Actions are quite small and just delegate the logic to the specific package. I find this easier to read.

File: integrationlink/params.go [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+func newParamsChecker(repoLink *scalingo.SCMRepoLink, ctx *cli.Context) *paramsChecker {
+	return &paramsChecker{repoLink: repoLink, ctx: ctx}
+}
+
+func (p *paramsChecker) lookupBranch() *string {
+	branch := p.ctx.String("branch")
+
+	if branch != "" && p.repoLink.Branch != branch {
```
Comment (johnsudaar) [inline integrationlink/params.go]: Why are those test needed? if we don't do them, we do not need to show the RepoLink which will remove the Show API call.

File: integrationlink/show.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+		return nil
+	}
+
+	i, err := c.SCMIntegrationsShow(repoLink.AuthIntegrationUUID)
+	if err != nil {
+		return errgo.Notef(err, "fail to get integration information of this integration link")
+	}
+
+	t := tablewriter.NewWriter(os.Stdout)
+	t.SetHeader([]string{
```
Comment (johnsudaar) [inline integrationlink/show.go]: A table for a single element ?

Why not using something that is a bit more readable like:

```
App: myapp (58ab56456def)
Integration: gitlab-self-hosted (scm-asd4556-5645-564)

Repo: Scalingo/api
Branch: prod
Owner: johnsudaar

AutoDeploy: true

Review Apps: Enabled
   Destroy on close: 1h
   Destroy on Stale: 2h
```

File: integrationlink/create.go [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+	if err != nil {
+		return errgo.Notef(err, "fail to get Scalingo client")
+	}
+
+	integration, err := c.SCMIntegrationsShow(string(integrationType))
+	if err != nil {
+		return errgo.Notef(err, "fail to get the integration")
+	}
+
+	repoLink, err := c.SCMRepoLinkShow(app)
```
Comment (johnsudaar) [inline integrationlink/create.go]: Wont SCMRepoLinkShow send an error if not SCM is configured (like a 404 ?)

File: cmd/scm_integrations.go [hunk 15]
```diff
@@ -1,167 +1,168 @@
...
 			if c.NArg() != 1 {
-				_ = cli.ShowCommandHelp(c, "scm-integrations-create")
+				_ = cli.ShowCommandHelp(c, "integrations-add")
 				return
 			}
 
-			link := c.String("url")
+			integrationURL := c.String("url")
 			token := c.String("token")
 			scmType := scalingo.SCMType(c.Args()[0])
```
Comment (johnsudaar) [inline cmd/scm_integrations.go]: Do we want to try to get the type from the URL?

File: integrationlink/show.go (hunk 13 above)
Comment (EtienneM) [reply integrationlink/show.go]: Screenshot from 2019-08-30 12-44-05

I used the same color as for the display of logs archive

File: integrationlink/create.go (hunk 14 above)
Comment (EtienneM) [reply integrationlink/create.go]: Surprisingly, no..

File: cmd/scm_integrations.go (hunk 15 above)
Comment (EtienneM) [reply cmd/scm_integrations.go]: Here you are in the `scm_integrations` file. So a user type one of the following commands:

```
scalingo integrations-add github
scalingo integrations-add gitlab
scalingo integrations-add github-enterprise --url https://ghe.example.com --token personal-access-token
scalingo integrations-add gitlab-self-hosted --url https://gitlab.example.com --token personal-access-token
```

How would you deduce the type from the URL?

Comment (johnsudaar) [reply cmd/scm_integrations.go]: I thought of `scalingo integrations-add --url https://github.com` but yeah it's not really good.

**Pull request merged**